### ISSUE TITLE
set `--augmentation 0` by default in `rasa train` and `rasa train core`

### ISSUE THREAD

**Issue by Ghostvv** [RasaHQ/rasa#6550]

**Comment by tabergma** [comment]
We need to test if changing the default value drops performance. To test this we need to wait for Sara and Carbon Bot to update to the latest Rasa 2.0 changes, e.g. they should use rules + TED. Until then the feature is on hold.

**Comment by Ghostvv** [comment]
we can also do experiments on current Sara and carbon-bot, but we need test set formed from real conversations (cc @akelad)

**Comment by Ghostvv** [comment]
@samsucik this is related to TED + rules

**Comment by samsucik** [comment]
Currently waiting for a [new Sara test set](https://github.com/RasaHQ/research/issues/115).

**Comment by Ghostvv** [comment]
I moved it because it is significantly bigger than sprint issue

**Comment by samsucik** [comment]
# First numbers are in, lots of work ahead

## What's been done
I have evaluated different policy combinations with/without story augmentation on the new Sara test set, numbers [here](https://docs.google.com/spreadsheets/d/1pOaWqoK8VQHa77Hs0CY4aU9_Qn98KJXQ47BDQtUR5iA/edit?usp=sharing). In short, `AugMemo` doesn't need story augmentation but `TED` benefits from it.

To better understand what's behind the numbers -- in particular, what mistakes `TED` makes that `AugMemo` doesn't -- I've started looking at the responses fir which the F1 difference is large between two policy combinations: `AugMemo+TED+Rules` (the default config for Sara) vs `AugMemo+Rules`. These mistakes point towards:
- issues in Sara's training data and occasionally also in the new test set
- issues with TED (so far, especially with TED's attention mechanism)

## Next steps
- [ ] try TED with relative attention; see if it eliminates mistakes that point toward "bag-of-states" treatment
- [ ] finish manual inspection of mistakes on the test set to see how mistakes vary across seeds
- [ ] use Johannes's attention visualisation tool to check/confirm some hypotheses about TED's mistakes
- [ ] look at mistakes on the training stories and try to match them with those on the test stories
- [ ] categorise the mistake types (by the likely cause) to create a more readable, actionable summary

**Comment by samsucik** [comment]
# Final update
In short, [the numbers](https://docs.google.com/spreadsheets/d/1pOaWqoK8VQHa77Hs0CY4aU9_Qn98KJXQ47BDQtUR5iA/edit?ts=5f53fe97&pli=1#gid=1823400684) show that TED benefits from story augmentation and making `0` the default might not be wise.

Further, this project has turned into an analysis of the mistakes that TED makes but AugMemo doesn't (i.e. what kinds of memorisation TED struggles with). The primary outcome is [this presentation](https://docs.google.com/presentation/d/19gaGKokVnB91FbfxsYume0lrAg_JTuSatFA3Cmfz_1k/edit?usp=sharing). Further exploration of why TED makes particular [kinds of] mistakes is out of the scope of this project and @JEM-Mosig is looking into it as part of [his TED investigation](https://github.com/RasaHQ/research/issues/134).

I'll create a more thorough summary of this project soon in our wiki.

**Comment by samsucik** [comment]
@akelad I'm think about how to wrap this up (at least for the foreseeable future). Do you think users would benefit from some new recommendations, or shall we just say that the main outcome is us internally keeping story augmentation as a default? Btw see my last comment above for the gist.

**Comment by akelad** [comment]
What would the new recommendations be?

**Comment by samsucik** [comment]
I'd say in the docs that `augmentation` significantly affects training times (we know this very well but don't articulate it in the docs) and that we recommend:
- to start with augmentation (whether the default `50` or even more) when the config contains TED and no memoization policies,
- trying also with augmentation turned off if a memoization policy is present as well (to potentially speed up training if numbers look good even without augmentation).

_I guess the fact that `augmentation` is a CLI parameter (rather than a config one) makes this a bit unusual/awkward (and I imagine that the parameter gets overlooked much more than config params), but we can't figure out everything at once -- and perhaps this will change in 3.0._

**Comment by akelad** [comment]
I'm not 100% sure this needs to be explained in any more detail in the docs - we have some guidance on how to use this flag right? I don't think that this issue came up from any external users, but rather something we wanted to try out internally.

**Comment by samsucik** [comment]
@akelad you now made me read the [relevant docs part](https://rasa.com/docs/rasa/policies#data-augmentation) carefully and it was pretty enlightening 🙂  I don't think we need any drastic changes. Definitely not suggesting that we _add lots of new info_, but I think we could improve the current info (and I'm happy to make the changes):
- Make it explicit that `augmentation` is a crucial parameter when trying to speed up training -- even if no-one asked specifically about augmentation, we definitely had people asking about Core training being slow, right?
- Correct the docs to state the correct default augmentation factor (they say it's 20 but it's 50).
- Make it explicit that augmentation affects only TED (out of the policies we provide)

What do you think?

**Comment by akelad** [comment]
yeah these sound like good changes, go for it!
